src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs

```diff
@@ -1064,304 +907,250 @@ internal TransitionRegex<S> CreateDerivative()
+        /// <summary>
+        /// Takes the derivative of the symbolic regex for the given element, which must be either
+        /// a minterm (i.e. a class of characters that have identical behavior for all predicates in the pattern)
+        /// or a singleton set. This derivative is "eager", i.e. it only considers paths that backtracking would
+        /// take before accepting the empty string for this pattern and returns the pattern ordered in the order backtracking
+        /// would explore paths. For example the derivative of a*ab for a is a*ab|b, while for a*?ab it is b|a*?ab.
+        /// </summary>
+        /// <param name="elem">given element wrt which the derivative is taken</param>
+        /// <param name="context">immediately surrounding character context that affects nullability of anchors</param>
+        /// <returns>the derivative</returns>
+        internal SymbolicRegexNode<S> CreateEagerDerivative(S elem, uint context)
+        {
+            List<(SymbolicRegexNode<S>, DerivativeEffect[]?)> transitions = new();
+            AddTransitions(elem, context, transitions, new List<SymbolicRegexNode<S>>(), null);
+            SymbolicRegexNode<S> derivative = _builder._nothing;
+            for (int i = transitions.Count - 1; i >= 0; --i)
+            {
+
+                derivative = _builder.OrderedOr(transitions[i].Item1, derivative);
+            }
+            return derivative;
+        }
 
         /// <summary>
-        /// Computes the symbolic derivative as a transition regex.
-        /// Transitions are in the tree left to right in the order the backtracking engine would explore them.
-        /// The transitions also include the effects (e.g. capture starts and ends) along their paths.
+        /// Takes the derivative of the symbolic regex for the given element, which must be either
+        /// a minterm (i.e. a class of characters that have identical behavior for all predicates in the pattern)
+        /// or a singleton set. This derivative is "eager", i.e. it only considers paths that backtracking would
+        /// take before accepting the empty string for this pattern and returns the pattern ordered in the order backtracking
+        /// would explore paths. For example the derivative of a*ab places a*ab before b, while for a*?ab the order is reversed.
         /// </summary>
-        /// <param name="eager">whether to only include paths before the first time the backtracking matchers would hit nullability</param>
-        /// <returns>the derivative as a TransitionRegex/></returns>
-        internal TransitionRegex<S> CreateDerivativeWithEffects(bool eager)
+        /// <remarks>
+        /// The differences of this to <see cref="CreateEagerDerivative"/> are that (1) effects (e.g. capture starts and ends)
+        /// are considered and (2) the different elements that would form a top level union are instead returned as separate
+        /// nodes (paired with their associated effects).
```

Comment:
Can you augment this to explain why we would want the top level union nodes separated in this case?